File: leetcode_30days/03042020_maximum_subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  var lmax = 0;
  var gmax = nums[0];

  for (var i = 0; i < nums.length; i++) {
    lmax = Math.max(nums[i], nums[i] + lmax);
    gmax = Math.max(lmax, gmax);
  }

  return gmax;
};

File: leetcode_30days/03042020_maximum_subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
  lmax = 0
  gmax = nums.first

  nums.each do |num|
    lmax = [num, num + lmax].max
    gmax = [lmax, gmax].max
  end

  gmax
end
